edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs

```diff
@@ -210,22 +217,22 @@ async Task<Option<ServiceIdentity>> GetServiceIdentityInternal(string id)
             Preconditions.CheckNonWhiteSpace(id, nameof(id));
             using (await this.cacheLock.LockAsync())
```

Comment:
do we still need lock here since you changed to use AtomicReference?

Reply:
I think you're right, but based on David's suggestion for removing IAuthenticationChainProvider, maybe we can get rid of the AtomicReference in ServiceIdentityTree, which means wel'l need this lock again.

---
In reply to: [428953358](https://github.com/Azure/iotedge/pull/2987#discussion_r428953358) [](ancestors = 428953358)

Reply:
The tree impl is now threadsafe via locks, and I removed the locking at the caching layer.